source/Octopus.Tentacle/Maintenance/WorkspaceCleaner.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Communications;
+using Octopus.Tentacle.Scripts;
+using Octopus.Tentacle.Services.Scripts;
+using Octopus.Time;
+
+namespace Octopus.Tentacle.Maintenance
+{
+    public class WorkspaceCleaner
+    {
+        readonly WorkspaceCleanerConfiguration configuration;
+        readonly IScriptWorkspaceFactory scriptWorkspaceFactory;
+        readonly IClock clock;
+        readonly ISystemLog log;
+
+        readonly ScriptService scriptService;
+        readonly ScriptServiceV2 scriptServiceV2;
+
+        public WorkspaceCleaner(
+            WorkspaceCleanerConfiguration configuration,
+            IScriptWorkspaceFactory scriptWorkspaceFactory,
+            IServiceRegistration serviceRegistration,
+            IClock clock,
+            ISystemLog log)
+        {
+            this.configuration = configuration;
+            this.scriptWorkspaceFactory = scriptWorkspaceFactory;
+            this.clock = clock;
+            this.log = log;
+
+            scriptService = serviceRegistration.GetService<ScriptService>();
+            scriptServiceV2 = serviceRegistration.GetService<ScriptServiceV2>();
+        }
+
+        public async Task Clean(CancellationToken cancellationToken)
+        {
+            var deleteWorkspacesOlderThanDateTimeUtc = clock.GetUtcTime().DateTime.Subtract(configuration.DeleteWorkspacesOlderThanTimeSpan);
+            log.Info($"Cleaning orphaned workspaces older than {deleteWorkspacesOlderThanDateTimeUtc:g}");
+
+            var deletedCount = 0;
+            var workspaces = scriptWorkspaceFactory.GetUncompletedWorkspaces();
+            foreach (var workspace in workspaces)
+            {
+                try
+                {
+                    if (scriptService.IsRunningScript(workspace.ScriptTicket)) continue;
+                    if (scriptServiceV2.IsRunningScript(workspace.ScriptTicket)) continue;
+
+                    var workspaceLogFilePath = workspace.LogFilePath;
+                    if (!File.Exists(workspaceLogFilePath)) continue;
+
+                    var outputLogFileLastWriteTimeUtc = File.GetLastWriteTimeUtc(workspaceLogFilePath);
+                    if (outputLogFileLastWriteTimeUtc >= deleteWorkspacesOlderThanDateTimeUtc) continue;
+
+                    await workspace.Delete(cancellationToken);
+                    deletedCount++;
+                }
+                catch (Exception e)
+                {
+                    log.Error(e, $"Could not delete orphaned workspace {workspace.WorkingDirectory}");
+                }
+            }
+
+            log.Info($"Deleted {deletedCount} orphaned workspace{(deletedCount != 1 ? "s" : "")}");
```

Comment:
Do you think it's useful to log we did nothing e.g. `Deleted 0 orphaned` at info Level or would it be better to suppress this log noise a bit e.g. just log at info when we did some work otherwise log at verbose etc.

Reply:
🤔 I don't have any strong opinions here. So I'll list some statements to help make a choice 😁 

This only runs every 2 hours (which we can make once a day?), so logging will be minimal.

Regardless, we are already logging at info level that we are going to attempt it. So it might be nice to see what the result was. I know Verbose is also being logged, but really, who looks at Verbose unless something goes wrong.

We could combine these. But having the log entry at the start is helpful to know if we aren't reaching it (plus helps show how long it took).

Personally, I'd either leave it, or combine both entries into 1 entry at the end.

But if you think verbose is superior, I can do that?

Reply:
> we are already logging at info level that we are going to attempt it

This could be a candidate for Verbose as well

E.g. how useful are these log messages to users?

I think it's useful to know when it did delete something (just incase this is the cause of a problem) or if there was an error when running the job - so these seem like Info or higher candidate.

We're going to run this thing every 2 hours and we did run it and it did nothing seem like good candidate for Verbose logging.

Reply:
Sounds good to me. To summarise:
- Make top log Verbose
- Make bottom log Verbose if 0 (maybe even a new message as well)
- Keep bottom log Info if != 0